craft_cli/messages.py

```diff
@@ -388,6 +393,92 @@ def advance(self, amount: Union[int, float]) -> None:
         self.printer.progress_bar(self.stream, self.text, self.accumulated, self.total)
 
 
+class _PipeReaderThread(threading.Thread):
+    """A thread that reads bytes from a pipe and write lines to the Printer."""
+
+    def __init__(self, pipe: int, printer: _Printer, stream: Optional[TextIO]):
+        super().__init__()
+
+        # the pipe to read from
+        self.read_pipe = pipe
+
+        # special flag used to stop the pipe reader thread
+        self.stop_flag = False
+
+        # where to collect the content that is being read but yeat not written (waiting for
+        # a newline)
+        self.remaining_content = b""
+
+        # printer and stream to write the assembled lines
+        self.printer = printer
+        self.stream = stream
+
+    def _write(self, data: bytes) -> None:
+        """Convert the byte stream into unicode lines and send it to the printer."""
+        pointer = 0
+        data = self.remaining_content + data
+        while True:
+            # get the position of next newline (find starts in pointer position)
+            newline_position = data.find(b"\n", pointer)
```

Comment:
Since data is a byte array containing characters encoded in utf-8, what happens if 0x0a is part of the code for a different unicode character?

Reply:
It can't. All UTF8 bytes start with `1`, except those unicode points from 0 to 127. See "Encoding" [here](https://en.wikipedia.org/wiki/UTF-8).